File: server/src/plugins/runtime/node/child.ts
import { Logger } from '@camera.ui/common/logger';
import { SignalHandler } from '@camera.ui/common/utils';
import { createRPCClient } from '@camera.ui/rpc';
import { API_EVENT } from '@camera.ui/sdk';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { isPromise } from 'node:util/types';

import { NamespaceManager } from '../../../rpc/namespaces.js';
import { PluginStoreFile } from '../../store/pluginStoreFile.js';
import { PLUGIN_COMMAND, PLUGIN_STATUS } from '../../types.js';
import { LocalPluginConfigDb, RemotePluginConfigDb } from './configDb.js';
import { PluginAPI } from './pluginApi.js';
import { CameraDeviceProxy } from './proxy/cameraDevice.js';
import { PluginFileServer } from './proxy/fileServe.js';

import type { LoggerOptions } from '@camera.ui/common';
import type { Channel, RPCClient } from '@camera.ui/rpc';
import type { BasePlugin, Camera, DeviceStorage, PluginInfo } from '@camera.ui/sdk';
import type { PluginConfigStoreRPC } from '../../config-store.js';
import type { ProcessLoadMessage, ProcessMessage, ProcessResponse } from '../../types.js';
import type { PluginConfigDb } from './configDb.js';

type PluginConstructor = new (logger: Logger, api: PluginAPI, storage: DeviceStorage<Record<string, any>>) => BasePlugin;

const SHUTDOWN_LISTENER_TIMEOUT = 1000;
const RPC_TEARDOWN_TIMEOUT = 500;

const processName = process.argv[3] || process.argv[2] || 'Plugin';
process.title = `camera.ui - ${processName}`;

export class PluginChild {
  private proxy: RPCClient;
  private channel?: Channel;

  private displayName: string;
  private api?: PluginAPI;
  private plugin?: BasePlugin;
  private closeProxy?: () => Promise<void>;
  private pluginDb?: PluginConfigDb;
  private fileServer?: PluginFileServer;
  private signalHandler: SignalHandler;

  private logger: Logger;
  private loggerOptions: LoggerOptions;

  private stopped = false;
  constructor() {
    this.proxy = createRPCClient({
      name: NamespaceManager.pluginNamespaces(process.env.PLUGIN_ID!).pluginChild,
      servers: process.env.PROXY_ENDPOINTS!.split(','),
      auth: {
        user: process.env.PROXY_USER!,
        password: process.env.PROXY_PASSWORD!,
      },
      // tls: {
      //   cert: process.env.PROXY_CERT!,
      //   key: process.env.PROXY_KEY!,
      //   ca: process.env.PROXY_CA!,
      // },
    });

    delete process.env.PROXY_USER;
    delete process.env.PROXY_PASSWORD;
    delete process.env.PROXY_ENDPOINTS;
    delete process.env.PROXY_CERT;
    delete process.env.PROXY_KEY;
    delete process.env.PROXY_CA;

    this.displayName = processName;

    this.loggerOptions = {
      prefix: this.displayName,
      debugEnabled: process.env.LOGGER_LEVEL === 'debug' || process.env.LOGGER_LEVEL === 'trace',
      traceEnabled: process.env.LOGGER_LEVEL === 'trace',
      targetId: process.env.PLUGIN_ID,
      targetType: 'plugin',
      pluginId: process.env.PLUGIN_ID,
    };

    this.logger = new Logger(this.loggerOptions);
    this.logger.setChildProcessMode(true);

    this.signalHandler = new SignalHandler({
      displayName: '[Signal]',
      timeoutDuration: 2000,
      logger: this.logger,
      closeFunction: this.stopPlugin.bind(this),
    });
  }

  public async run(): Promise<void> {
    try {
      await this.proxy.connect();
      await this.onStart();
    } catch (error: any) {
      this.logger.error(`Failed to connect to proxy server: ${error.message}`);
      process.exit(1);
    }
  }

  private async onStart(): Promise<void> {
    this.channel = await this.proxy.privateChannel('plugin-communication', 'camera.ui');
    this.channel.on('message', this.onMessage.bind(this));
    await this.sendMessage({ type: PLUGIN_STATUS.READY });
  }

  private async startPlugin(data: ProcessLoadMessage): Promise<void> {
    try {
      // Host-local writable dir — on a remote worker the master's path from
      // the START message would point at the wrong machine.
      const storagePath = process.env.PLUGIN_STORAGE_PATH ?? data.storage.storagePath;
      const pluginInfo = data.plugin;
      const cameras = data.cameras;

      this.pluginDb = await this.configurePluginDB(storagePath, pluginInfo.id);
      this.api = new PluginAPI(this.proxy, pluginInfo, storagePath, this.logger, this.pluginDb);

      const PluginConstructor = await this.loadPlugin();
      const pluginStorage = await this.api._storageController.createStorage('plugin');
      this.plugin = new PluginConstructor(this.logger, this.api, pluginStorage);

      if (this.plugin.storageSchema && this.plugin.storageSchema.length > 0) {
        pluginStorage.defineSchemas(this.plugin.storageSchema);
      }

      const namespaces = NamespaceManager.pluginNamespaces(pluginInfo.id);
      this.closeProxy = await this.proxy.registerHandler(namespaces.pluginChildRpc, this.plugin, { withoutDecorators: true });

      // Remote-hosted: serve this worker's files so the master can stream
      // downloads/exports of them.
      if (process.env.PLUGIN_REMOTE_MODE) {
        this.fileServer = new PluginFileServer(this.proxy, pluginInfo.id);
        await this.fileServer.register();
      }

      this.api.deviceManager.setPlugin(this.plugin);

      await this.api.deviceManager.init();
      await this.api.coreManager.init();
      await this.configureCameras(this.api, pluginInfo, cameras);

      this.sendMessage({ type: PLUGIN_STATUS.STARTED });

      setImmediate(() => {
        this.api?.emit(API_EVENT.FINISH_LAUNCHING);
      });
    } catch (error: any) {
      this.sendMessage({ type: PLUGIN_STATUS.ERROR, error: error.message });
    }
  }

  private async stopPlugin(): Promise<void> {
    if (this.stopped) {
      return;
    }

    this.stopped = true;

    await this.emitShutdownAndWait();
    await this.api?.deviceManager.close();
    await this.api?.coreManager.close();
    await this.api?._storageController.close();
    await this.pluginDb?.close?.();

    const teardown = (async () => {
      await this.fileServer?.close();
      await this.channel?.close();
      await this.closeProxy?.();
      await this.proxy.disconnect();
    })();

    let timer: NodeJS.Timeout | undefined;
    const timedOut = await Promise.race([
      teardown.then(
        () => false,
        () => false,
      ),
      new Promise<boolean>((resolve) => {
        timer = setTimeout(() => resolve(true), RPC_TEARDOWN_TIMEOUT);
      }),
    ]);
    clearTimeout(timer);

    if (timedOut) {
      this.logger.warn(`RPC teardown still pending after ${RPC_TEARDOWN_TIMEOUT}ms, force-closing transport`);
      this.proxy.abortClose();
    }
  }

  private async sendMessage(message: ProcessResponse): Promise<void> {
    await this.channel?.send<ProcessResponse>(message);
  }

  private async onMessage(message: ProcessMessage): Promise<void> {
    if (this.stopped) {
      return;
    }

    switch (message.type) {
      case PLUGIN_COMMAND.START:
        if (!message.data) {
          this.sendMessage({ type: PLUGIN_STATUS.ERROR, error: 'No data provided' });
          return;
        }

        this.startPlugin(message.data);
        break;
      case PLUGIN_COMMAND.STOP:
        this.stopPlugin();
        break;
    }
  }

  private async configurePluginDB(storagePath: string, pluginId: string): Promise<PluginConfigDb> {
    // Remote-hosted: config lives on the master (re-homing safe) — persist
    // through its config store instead of a worker-local file.
    if (process.env.PLUGIN_CONFIG_STORE_RPC) {
      const namespaces = NamespaceManager.pluginNamespaces(pluginId);
      const remoteDb = new RemotePluginConfigDb(this.proxy.createProxy<PluginConfigStoreRPC>(namespaces.pluginConfigStoreRpc));
      await remoteDb.init();
      return remoteDb;
    }

    const store = new PluginStoreFile(`${storagePath}/volume`, pluginId);
    await store.open();
    return new LocalPluginConfigDb(store);
  }

  private async configureCameras(api: PluginAPI, pluginInfo: PluginInfo, cameras: Camera[]): Promise<void> {
    const cameraDevices = cameras.map((camera) => {
      const cameraLogger = this.logger.createLogger({ suffix: camera.name, targetId: camera._id, targetType: 'camera' });
      return new CameraDeviceProxy(this.proxy, api._storageController, camera, pluginInfo, cameraLogger);
    });

    await this.api?.deviceManager.configureCameras(cameraDevices);
    await this.plugin?.configureCameras(cameraDevices);
  }

  private async loadPlugin(): Promise<PluginConstructor> {
    const moduleUrl = pathToFileURL(process.env.MODULE_PATH!).href;
    const mod = await import(moduleUrl);
    const isDefaultPromise = isPromise(mod.default);

    let defaultModule: any;

    if (isDefaultPromise) {
      defaultModule = await mod.default;
    } else {
      defaultModule = mod.default;
    }

    if (defaultModule && typeof defaultModule.default === 'function') {
      return defaultModule.default;
    } else if (typeof defaultModule === 'function') {
      return defaultModule;
    } else {
      throw new Error(`Plugin ${process.env.MODULE_PATH} does not export a initializer function from main.`);
    }
  }

  private async emitShutdownAndWait(): Promise<void> {
    if (!this.api) {
      return;
    }

    // Detach before invoking so once-wrappers cannot fire a second time.
    const listeners = this.api.rawListeners(API_EVENT.SHUTDOWN) as ((...args: unknown[]) => unknown)[];
    this.api.removeAllListeners(API_EVENT.SHUTDOWN);

    if (listeners.length === 0) {
      return;
    }

    const settled = Promise.all(
      listeners.map(async (listener) => {
        try {
          await listener();
        } catch (error: any) {
          this.logger.error(`Shutdown listener failed: ${error.message}`);
        }
      }),
    );

    let timer: NodeJS.Timeout | undefined;
    const timedOut = await Promise.race([
      settled.then(() => false),
      new Promise<boolean>((resolve) => {
        timer = setTimeout(() => resolve(true), SHUTDOWN_LISTENER_TIMEOUT);
      }),
    ]);
    clearTimeout(timer);

    if (timedOut) {
      this.logger.warn(`Shutdown listeners still pending after ${SHUTDOWN_LISTENER_TIMEOUT}ms, continuing teardown`);
    }
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const plugin = new PluginChild();
  await plugin.run();
}

File: server/src/plugins/runtime/python/child.py
from __future__ import annotations

import truststore

truststore.inject_into_ssl()

import asyncio
import importlib.util
import os
import sys
from contextlib import suppress
from pathlib import Path
from types import ModuleType
from typing import Any, cast

import setproctitle

from _camera_ui_tools.camera_ui_common import (
    LoggerOptions,
    LoggerService,
    SignalHandler,
)
from _camera_ui_tools.camera_ui_rpc import (
    CloseHandler,
    PrivateChannel,
    RPCClient,
    create_rpc_client,
)
from _camera_ui_tools.camera_ui_sdk import API_EVENT, BasePlugin, Camera
from _camera_ui_tools.camera_ui_sdk import CameraDevice as CameraDeviceInterface
from plugins.runtime.python.config_db import (
    LocalPluginConfigDb,
    PluginConfigDb,
    PluginConfigStoreRPC,
    RemotePluginConfigDb,
)
from plugins.runtime.python.namespaces import NamespaceManager
from plugins.runtime.python.plugin_api import PluginAPI
from plugins.runtime.python.proxy.camera_device import CameraDeviceProxy
from plugins.runtime.python.proxy.file_serve import PluginFileServer
from plugins.runtime.python.store import PluginStoreFile
from plugins.runtime.python.typings import (
    PLUGIN_COMMAND,
    PLUGIN_STATUS,
    PluginInfo,
    ProcessLoadMessage,
    ProcessMessage,
    ProcessResponse,
)

process_name = sys.argv[2] if len(sys.argv) > 2 else sys.argv[1] if len(sys.argv) > 1 else "Plugin"
setproctitle.setproctitle(f"camera.ui - {process_name}")

SHUTDOWN_LISTENER_TIMEOUT = 1.5
RPC_TEARDOWN_TIMEOUT = 1.0
CONNECT_DEADLINE = 15.0


class PluginChild:
    def __init__(self) -> None:
        self.endpoints = os.environ["PROXY_ENDPOINTS"].split(",")
        self._connect_error: str | None = None
        self._connect_failed = asyncio.Event()

        self.proxy: RPCClient = create_rpc_client(
            {
                "name": NamespaceManager.plugin_namespaces(os.environ["PLUGIN_ID"]).plugin_child,
                "servers": self.endpoints,
                "auth": {
                    "user": os.environ["PROXY_USER"],
                    "password": os.environ["PROXY_PASSWORD"],
                },
                "reconnected_cb": self._on_rpc_reconnected,
                "error_cb": self._on_rpc_error,
            }
        )

        for key in [
            "PROXY_USER",
            "PROXY_PASSWORD",
            "PROXY_ENDPOINTS",
            "PROXY_CERT",
            "PROXY_KEY",
            "PROXY_CA",
        ]:
            if key in os.environ:
                del os.environ[key]

        self.display_name = process_name

        logger_options: LoggerOptions = {
            "prefix": self.display_name,
            "debug_enabled": os.environ.get("LOGGER_LEVEL") in ["debug", "trace"],
            "trace_enabled": os.environ.get("LOGGER_LEVEL") == "trace",
            "target_id": os.environ.get("PLUGIN_ID"),
            "target_type": "plugin",
            "plugin_id": os.environ.get("PLUGIN_ID"),
        }

        self.logger = LoggerService(logger_options)
        self.logger.set_child_process_mode(True)

        self.api: PluginAPI | None = None
        self.plugin: BasePlugin | None = None
        self.channel: PrivateChannel | None = None
        self.close_proxy: CloseHandler | None = None
        self.file_server: PluginFileServer | None = None
        self.plugin_db: PluginConfigDb | None = None

        self.stopped = False

        self.signal_handler = SignalHandler(
            {
                "display_name": "[Signal]",
                "timeout_duration": 2,
                "logger": self.logger,
                "close_function": self.stop_plugin,
            }
        )

    async def run(self) -> None:
        try:
            self.signal_handler.setup_handlers()

            try:
                await self.connect_proxy()
            except Exception as error:
                self.logger.error(str(error))
                raise SystemExit(1) from None

            await self.on_start()
            await self.signal_handler.shutdown_event.wait()
        except asyncio.CancelledError:
            pass
        except Exception as e:
            self.logger.error(f"Failed to connect to proxy server: {e}")
            raise
        finally:
            await self.stop_plugin()

    async def connect_proxy(self) -> None:
        connecting = asyncio.ensure_future(self.proxy.connect())
        failed = asyncio.ensure_future(self._connect_failed.wait())

        try:
            await asyncio.wait(
                {connecting, failed},
                timeout=CONNECT_DEADLINE,
                return_when=asyncio.FIRST_COMPLETED,
            )

            if connecting.done():
                await connecting
                return

            reason = self._connect_error or f"no answer within {CONNECT_DEADLINE:.0f}s"
            raise RuntimeError(f"camera.ui not reachable on {', '.join(self.endpoints)}: {reason}")
        finally:
            for task in (connecting, failed):
                task.cancel()

    async def _on_rpc_error(self, error: Exception) -> None:
        self._connect_error = f"{type(error).__name__}: {error}" if str(error) else type(error).__name__
        self._connect_failed.set()

    async def _on_rpc_reconnected(self) -> None:
        if self.stopped or self.api is None:
            return
        self.api.sensor_manager.on_rpc_reconnected()
        self.api.device_manager.on_rpc_reconnected()

    async def on_start(self) -> None:
        self.channel = await self.proxy.private_channel("plugin-communication", "camera.ui")
        self.channel.on("message", self.on_message)
        await self.send_message({"type": PLUGIN_STATUS.READY.value})

    async def start_plugin(self, data: ProcessLoadMessage) -> None:
        try:
            # Host-local writable dir — on a remote worker the master's path from
            # the START message would point at the wrong machine.
            storage_path = os.environ.get("PLUGIN_STORAGE_PATH") or data["storage"]["storagePath"]
            plugin_info = data["plugin"]
            cameras = data["cameras"]

            self.plugin_db = await self.configure_plugin_db(storage_path, plugin_info["id"])

            self.api = PluginAPI(
                self.proxy,
                plugin_info,
                storage_path,
                self.logger,
                self.plugin_db,
            )

            plugin_constructor = self.load_plugin()
            plugin_storage = await self.api.storage_controller.createStorage("plugin")
            self.plugin = cast(BasePlugin, plugin_constructor(self.logger, self.api, plugin_storage))

            if hasattr(self.plugin, "storage_schema") and self.plugin.storage_schema:
                plugin_storage.defineSchemas(self.plugin.storage_schema)

            namespaces = NamespaceManager.plugin_namespaces(plugin_info["id"])
            self.close_proxy = await self.proxy.register_handler(
                namespaces.plugin_child_rpc, self.plugin, without_decorators=True
            )

            # Remote-hosted: serve this worker's files so the master can stream
            # downloads/exports of them.
            if os.environ.get("PLUGIN_REMOTE_MODE"):
                self.file_server = PluginFileServer(self.proxy, plugin_info["id"])
                await self.file_server.register()

            self.api.device_manager.set_plugin(self.plugin)
            self.api.sensor_manager.set_plugin(self.plugin)

            await self.api.device_manager.init()
            await self.api.core_manager.init()
            await self.configure_cameras(self.api, plugin_info, cameras)
            await self.api.sensor_manager.init()
            await self.api.sensor_manager.configure_adopted_sensors()

            await self.send_message({"type": PLUGIN_STATUS.STARTED.value})

            await asyncio.sleep(0.1)

            self.api.emit(API_EVENT.FINISH_LAUNCHING)
        except Exception as e:
            await self.send_message({"type": PLUGIN_STATUS.ERROR.value, "error": str(e)})

    async def stop_plugin(self) -> None:
        if self.stopped:
            return

        self.stopped = True

        if self.api:
            completed = await self.api.emit_and_wait(
                API_EVENT.SHUTDOWN.value,
                timeout=SHUTDOWN_LISTENER_TIMEOUT,
                on_error=lambda e: self.logger.error(f"Shutdown listener failed: {e}"),
            )
            if not completed:
                self.logger.warn(
                    f"Shutdown listeners still pending after {SHUTDOWN_LISTENER_TIMEOUT}s, continuing teardown"
                )

            await self.api.sensor_manager.close()
            await self.api.device_manager.close()
            await self.api.core_manager.close()
            await self.api.storage_controller.close()

            self.api.cancel()

        if self.plugin_db is not None:
            with suppress(Exception):
                await self.plugin_db.close()
            self.plugin_db = None

        try:
            await asyncio.wait_for(self._close_rpc(), timeout=RPC_TEARDOWN_TIMEOUT)
        except TimeoutError:
            self.logger.warn(f"RPC teardown still pending after {RPC_TEARDOWN_TIMEOUT}s, exiting anyway")

    async def _close_rpc(self) -> None:
        if self.channel:
            with suppress(Exception):
                await self.channel.close()

        if self.file_server:
            with suppress(Exception):
                await self.file_server.close()
            self.file_server = None

        if self.close_proxy:
            with suppress(Exception):
                await self.close_proxy()

        with suppress(Exception):
            await self.proxy.disconnect()

    async def send_message(self, message: ProcessResponse) -> None:
        if not self.channel:
            return

        await self.channel.send(message)

    async def on_message(self, message: ProcessMessage) -> None:
        if self.stopped:
            return

        match message["type"]:
            case PLUGIN_COMMAND.START.value:
                if not message["data"]:
                    await self.send_message({"type": PLUGIN_STATUS.ERROR.value, "error": "No data provided"})
                    return

                await self.start_plugin(message["data"])
            case PLUGIN_COMMAND.STOP.value:
                await self.stop_plugin()
            case _:
                pass

    async def configure_plugin_db(self, storage_path: str, plugin_id: str) -> PluginConfigDb:
        # Remote-hosted: config lives on the master (re-homing safe) — persist
        # through its config store instead of a worker-local file.
        if os.environ.get("PLUGIN_CONFIG_STORE_RPC"):
            namespaces = NamespaceManager.plugin_namespaces(plugin_id)
            store_rpc = self.proxy.create_proxy(namespaces.plugin_config_store_rpc, PluginConfigStoreRPC)
            remote_db = RemotePluginConfigDb(store_rpc)
            await remote_db.init()
            return remote_db

        volume_dir = Path(storage_path) / "volume"

        store = PluginStoreFile(str(volume_dir), self.logger)
        await store.open()
        return LocalPluginConfigDb(store)

    async def configure_cameras(self, api: PluginAPI, plugin_info: PluginInfo, cameras: list[Camera]) -> None:
        cameras_devices: list[CameraDeviceProxy] = []

        for camera in cameras:
            camera_logger = self.logger.create_logger(
                {
                    "suffix": camera["name"],
                    "target_id": camera["_id"],
                    "target_type": "camera",
                }
            )
            cameras_devices.append(
                CameraDeviceProxy(
                    self.proxy,
                    api.storage_controller,
                    api.sensor_manager,
                    camera,
                    plugin_info,
                    camera_logger,
                )
            )

        if self.api:
            await self.api.device_manager.configureCameras(cameras_devices)

        if self.plugin:
            await self.plugin.configureCameras(cast(list[CameraDeviceInterface], cameras_devices))

    def load_plugin(self) -> Any:
        module_path = os.environ.get("MODULE_PATH")
        if not module_path:
            raise Exception("MODULE_PATH environment variable not set")

        if not os.path.exists(module_path):
            raise Exception(f"Module path does not exist: {module_path}")

        spec = importlib.util.spec_from_file_location("plugin_module", module_path)
        if spec is None:
            raise Exception(f"Could not load module from '{module_path}'.")

        module = importlib.util.module_from_spec(spec)
        sys.modules["plugin_module"] = module

        if spec.loader is None:
            raise Exception(f"Could not load module from '{module_path}'.")

        spec.loader.exec_module(module)
        return self.find_plugin_constructor(module)

    def find_plugin_constructor(self, module: ModuleType) -> Any:
        if hasattr(module, "__main__"):
            main_func = module.__main__
            if callable(main_func):
                plugin_base_class = main_func()
                if isinstance(plugin_base_class, type):
                    return plugin_base_class
                else:
                    raise Exception("Plugin's __main__ must return a plugin constructor.")
            else:
                raise Exception("Plugin's __main__ is not callable.")

        else:
            raise Exception("Plugin does not have a __main__ function.")


if __name__ == "__main__":
    plugin_child = PluginChild()
    asyncio.run(plugin_child.run())
